### ISSUE TITLE
[css-layout-api] Generator vs. Promise design for the API

### ISSUE THREAD

**Issue by bfgeek** [w3c/css-houdini-drafts#750]
There isn't actually a large difference between the two APIs from just looking at the script, i.e.

```js
  *layout(children) {
    const fragment = yield children[0].layoutNextFragment();
  }
```

```js
  layout(children) {
    const fragment = await children[0].layoutNextFragment();
  }
```

There are various pros/cons of each of these.

Generator:
 - Pro: May be faster wrt. bindings.
 - Pro: More restrictive API, can't `await` on other APIs which might leak into the LWGS.
 - Con: Less obvious error handling to for web developers.
 - Con: Less obvious API as this pattern isn't common, and not many people use them (generators).

Promise:
 - Pro: More obvious error handling.
 - Pro: Common pattern for developers.

**Comment by tabatkins** [comment]
(Note that the resolution was explicitly meant to require discussion, not to shut down the debate and settle on one or the other.)

**Comment by dbaron** [comment]
A colleague noted the relationship of what's being done conceptually here to [coroutines](https://en.wikipedia.org/wiki/Coroutine), although I'm not sure if that yields anything useful.  (I vaguely recall that Ian might have mentioned that at the face-to-face meeting as well, though I don't see it minuted.)

**Comment by tabatkins** [comment]
Yeah, JS's iterator/yield and async/await are basically one-sided coroutines.
